### PR TITLE
fix: CdpCDPSession.connection() to return undefined after onClosed

### PR DESCRIPTION
**Pull request by alexey-pelykh**
**What kind of change does this PR introduce?**

Fix

**Did you add tests for your changes?**

~The tests should not be impacted.~
Yes

**If relevant, did you update the documentation?**

N/A

**Summary**

Restores `CdpCDPSession.connection()` behavior to return `undefined` after `onClosed` as it was before #13591

**Does this PR introduce a breaking change?**

It fix-reverts one

### CODE REVIEW

**Comment by OrKoN** [comment]
Thanks for the PR.

> Did you add tests for your changes?

> The tests should not be impacted.

Could you please add a test to `test/src/cdp/CDPSession.spec.ts`?:

```
  it('should not return a connection if the session is disposed', async () => {
    const {page} = await getTestState();
    const client = await page.createCDPSession();
    expect(client.connection()).toBeTruthy();
    await client.detach();
    expect(client.connection()).toBeUndefined();
    expect(Connection.fromSession(client)).toBeUndefined();
  });
```

**Comment by alexey-pelykh** [comment]
@OrKoN done

**Comment by Lightning00Blade** [comment]
I believe this change reintroduces the leaking SetTimeout that was fix with the change - #13591.

**Comment by alexey-pelykh** [comment]
@Lightning00Blade it may be so - however the original fix relied on exposing connection after detaching?

**Comment by Lightning00Blade** [comment]
Yes as we need to know if the whole browser is disconnecting or not - 
https://github.com/puppeteer/puppeteer/commit/6b20ac10b1cc6a54a87eb2f6437727d204acd1c3#diff-8ce6f7b6f62b4b2ea843fb6ed83de580cc083f4005adbc5c1671fbfe73792c98R107
I think the change made here is reasonable, as the behavior did change unintentionally.
Though I would am still interested in understanding the use case, as user is expected to call `.detach()` and and at that point they should not use the client. 

As for how to fix, I think introducing a internal `isConnectionClosed` and get that in the line above to use that.
Happy to create a PR after this as well.

**Comment by alexey-pelykh** [comment]
@Lightning00Blade thanks for providing the context - very helpful. I guess given than `connection()` is not marked internal (or explicitly public for that matter) - indeed it would be better to define the contract how it behaves.

The use-case (if I got you right): relying on `connection()` becoming `undefined` as an indicator that session is detached when it's not the callee who initiated the detachment.

**Comment by alexey-pelykh** [comment]
I guess the route would be either #13615 or internal `isConnectionClosed` as suggested @Lightning00Blade

**Comment by Lightning00Blade** [comment]
So the function I linked above handles when an internal CDPSessions gets detached, so at this point the `CDPSession.#closed` will be `true`, but the connection may be in either of the states, and we are interested if the connection is closed or not.

This is is roughly what happens:
Session 1 gets detached
Then there are two paths -> A) a new session will get attached (Event is emitted).
                                              -> B) the page is closed, no new session is attached.
So we set a check for an event and race it with a setTimeout(). 

But if the underlying connection is disconnected we know only B is possible, so we can jump to the code path for B.

I hope it makes sense.

**Comment by alexey-pelykh** [comment]
Thanks for the explanation! IMHO, then the `connection()` indeed should always return the underlying connection however a detached state should also be exposed as done in #13615

**Comment by alexey-pelykh** [comment]
At the same time, should there be a maintainer decision to have `connection()` behaving like before, I've updated this PR as well.

**Comment by alexey-pelykh** [comment]
Superseded by #13615
